doc/pcv.1
.\" Automatically generated by Pod::Man 2.1801 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PCV 1"
.TH PCV 1 "2008-12-21" "perl v5.10.0" "Picviz Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
pcv \- Picviz console veritable tool
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBpcv\fR \fB\-T\fRoutput_plugin [\fB\-R\fRrender_plugin] [\fBoptions\fR] file.pcv ['filter']
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBpcv\fR allows to compile \s-1PCV\s0 files into a plugin choosen format, such as \fBsvg\fR, \fBcsv\fR, \fBpng\fR
or other.
.PP
\&\fB\-Tplugin\fR Replace '\fBplugin\fR' with a output plugin name. See plugin section
.PP
\&\fB\-Rplugin\fR Replace '\fBplugin\fR' with a render plugin name. See plugin section
.PP
\&\fBoptions\fR can be either one or several among:
.PP
\&\fB\-A argument\fR Provides arguments to the render and output plugin
.PP
\&\fB\-a\fR Displays \fBall\fR text along with lines
.PP
\&\fB\-d\fR Activates debug mode
.PP
\&\fB\-Ln\fR Draw text every n line
.PP
\&\fB\-l\fR Skip learning mode
.PP
\&\fB\-o file\fR Output to the given file instead of stdout
.PP
\&\fB\-p file\fR Use given file to store \s-1PID\s0
.PP
\&\fB\-r...(rrr)\fR Increases image height and width
.PP
\&\fB\-s socket.name\fR Create the socket.name and listen to it (for real-time capabilities)
.PP
\&\fB\-t template\fR Use the given template when listening to a socket
.PP
\&\fB\-Warg\fR Use Picviz \fBwith\fR a special keyword. See keywords section
.PP
The filter parameter is explained in the filter section below.
.SH "PLUGINS"
.IX Header "PLUGINS"
Plugins are located with ld, search path can be overridden with \s-1PICVIZ_PLUGINS_PATH\s0 environment
variable.
.SS "Output"
.IX Subsection "Output"
Output data in the wanted file format. Available plugins are:
.PP
\&\fBsvg\fR
.PP
\&\fBplplot\fR
.PP
\&\fBcsv\fR
.PP
\&\fBsdl\fR
.PP
\&\fBpngcairo\fR
.PP
\&\fBdebug\fR
.SS "Render"
.IX Subsection "Render"
Modify data rendering. Available plugins are:
.PP
\&\fBheatline\fR: possible arguments (\fB\-A\fR): virus
.PP
\&\fBdebug\fR
.SH "KEYWORDS"
.IX Header "KEYWORDS"
Keywords are special parameters given to picviz to use a library along with parsing and rendering. This is
deactivated by default because it makes the program slow, but it may be very usefull in certain cases.
.PP
Following keywords are accepted:
.PP
\&\fBpcre\fR: activates pcre pattern matching for filtering (pcv \-Tsvg \-Wpcre file.pcv 'show value = \*(L".*foo.*\*(R" on axis 1')
.SH "FILTER"
.IX Header "FILTER"
pcv allows filtering to choose data you want to see displayed. When the pcv language parser is called,
a Pcv Image Filter (\s-1PIF\s0) is applied. It allows to filter either before any data is calculated (\fBdata filter\fR),
once calculated and before adding the line to the image (\fBpre line filter\fR) or, once other lines have been added
and ask for specific removales (\fBpost line filter\fR).
.SS "Relations"
.IX Subsection "Relations"
Filter relation can be either:
.PP
= equal
.PP
!= not equal
.PP
< less than
.PP
> greater than
.PP
<= less or equal
.PP
>= greater or equal
.SS "Selecting plot value"
.IX Subsection "Selecting plot value"
\&\fBplot\fR: Will select the value, as plotted by the engine. Accepted values can be the y position on the axis or
the relative percentage.
.PP
\&'show plot > 100 on axis 3': will display only lines which have their plot on the axis 3 greater than 100
can be called like this:
.PP
Example: pcv \-Tsvg file.pcv 'show plot > 100 on axis 3'
.PP
By default, the image height is 500 pixels, so this filter is equivalent than > 100:
.PP
\&'show plot > 20% on axis 3'.
.PP
It is possible to filter using multi-criterion values:
.PP
\&'show plot > 20% on axis 3 and plot < 42 on axis 5'
.SS "Pattern matching"
.IX Subsection "Pattern matching"
Values can be selected, either with their original text string, or by using pattern matching. By default,
pattern matching is \fBnot\fR activated.
.PP
Pattern matching activated is done with \fB\-Wpcre\fR argument. To filter values with \fB[0\-9][aA].*\fR, you can type:
.PP
pcv \-Tsvg file.pcv \-Wpcre 'show value = \*(L"[0\-9][aA].*\*(R" on axis 1'.
.PP
Note that with string comparisons (with \fBvalue\fR), only the equal (=) relation is accepted.
.SH "EXAMPLE"
.IX Header "EXAMPLE"
\&\fBpcv\fR \-Tsvg file.pcv
.PP
\&\fBpcv\fR \-Tpngcairo file.pcv \-rrrr \-a > file.png
.PP
\&\fBpcv\fR \-Tpngcairo \-Rheatline \-Avirus file.pcv > file.svg
.PP
\&\fBpcv\fR \-Tpngcairo \-s local.sock \-t file.pcv \-o file.png ''
.SH "BUGS"
.IX Header "BUGS"
If you are sure you are not dealing with a feature, you can report \s-1BUGS\s0 using the trac ticketing system
available at \fBhttp://www.wallinfire.net/picviz\fR.
.SH "AUTHORS"
.IX Header "AUTHORS"
Man page written by Sebastien Tricaud \fBtoady gscore org\fR.


src/pcv.c
/*
 * Picviz - Parallel coordinates ploter
 * Copyright (C) 2008 Sebastien Tricaud <toady@gscore.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation version 3.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * $Id: pcv.c 683 2009-07-20 22:02:37Z toady $
 */

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <sys/types.h>
#include <errno.h>

#include <picviz.h>

#define PCV_VERSION "0.6 ($Id: pcv.c 683 2009-07-20 22:02:37Z toady $)"

/* char debug = 0; */
char *plugin_arg = NULL;
char plugin_out_str[1024];
char plugin_ren_in = 0;
char plugin_ren_str[1024];
char *listen_sock = NULL;

/*
 * Copy arg vector into a new buffer, concatenating arguments with spaces.
 * stolen from tcpdump
 */
char *
concat_argv(char **argv)
{
	char **p;
	u_int len = 0;
	char *buf;
	char *src, *dst;

	p = argv;
	if (*p == 0)
		return 0;

	picviz_debug(PICVIZ_DEBUG_NOTICE, PICVIZ_AREA_FILTER, "Applying filter '%s'\n", *p);

	while (*p)
		len += strlen(*p++) + 1;

	buf = (char *)malloc(len);
	if (buf == NULL) {
		fprintf(stderr, "concat_argv: malloc");
		return NULL;
	}

	p = argv;
	dst = buf;
	while ((src = *p++) != NULL) {
		while ((*dst++ = *src++) != '\0')
			;
		dst[-1] = ' ';
	}
	dst[-1] = '\0';

	return buf;
}

void picviz_handlesig()
{
	fprintf(stderr, "[+] Terminating\n");
	if (listen_sock) {
		unlink(listen_sock);
	}
	unlink(engine.pid_file);
	exit(0);
}

void image_write_callback(PicvizImage *image)
{
	fprintf(stderr, ".");
	/* if (debug) { */
	/* 	picviz_image_debug_printall(image); */
	/* } */

	picviz_render_image(image); /* Our very costly function ;) */
	picviz_plugin_load(PICVIZ_PLUGIN_OUTPUT, plugin_out_str, image, plugin_arg);
}

void help(char *prg)
{
	fprintf(stderr, "Usage: %s -Tplugin [-Rplugin] [-Warg] [-ad] [-A argument] file.pcv ['filter']\nread the manpage pcv(1) for details\n",
			prg);
	exit(EXIT_FAILURE);
}

/* Write the pid number to the file 'file' */
void write_pid(char *file)
{
	FILE *fp;
	pid_t pid;

	pid = getpid();
	fp = fopen(file,"w");
	if (!fp) {
		if (errno) {
			fprintf(stderr, "[+] Using .picviz.pid to store PID\n");
			engine.pid_file = ".picviz.pid";
			fp = fopen(engine.pid_file,"w");
		} else {
			fprintf(stderr, "[E] Cannot write pid file '%s': %s\n", file, strerror(errno));
			fprintf(stderr, "[+] Terminating\n");
			exit(1);
		}
	}
	fprintf(fp,"%d",pid);
	fclose(fp);
}

int main(int argc, char **argv)
{
	struct pcimage_t *image;
	char *filename;
	char *filter = NULL;
	char plugin_in = 0;
	char draw_heatline = 0;
	int opt;
	unsigned int height = 0;
	unsigned int axis_default_space = 0;
	unsigned int axis_initial_cursor = 0;
	char *template = NULL; /* Template file for daemon mode */

	picviz_engine_init();

	signal(SIGTERM, picviz_handlesig);
	signal(SIGINT, picviz_handlesig);
	signal(SIGQUIT, picviz_handlesig);

	while ((opt = getopt(argc, argv, "L:lrT:R:W:dao:A:s:t:vVp:mq")) != -1) {
		switch (opt) {
			case 'A': /* Gives the argument to the plugin */
				plugin_arg = optarg;
				break;
			case 'a': /* Displays text */
				engine.display_raw_data = 1;
				break;
			case 'd': /* Starts in debug mode*/
				if (engine.debug < 255) {
					engine.debug += 1;
				}
				break;
			case 'L': /* Draw the text every N lines */
				engine.draw_text = atoi(optarg);
				break;
			case 'l': /* We don't learn the string algo */
				engine.learn = 0;
				break;
		        case 'm':
			        engine.display_minmax = 1;
				/* If we display min/max only, this is because we want text */
				engine.display_raw_data = 1;
			        break;
			case 'o': /* Output to a file instead of stdout */
				engine.output_file = optarg;
				break;
			case 'p':
				engine.pid_file = optarg;
				break;
			case 'q':
				engine.quiet = 1;
				break;
			case 'R': /* Rendering plugin (heatline, ...) */
				sprintf(plugin_ren_str, "libpicvizren%s.so", optarg);
				if ( ! strcmp("healine", optarg) ) {
					draw_heatline = 1;
				}
				plugin_ren_in = 1;
				break;
			case 'r': /* Inscreases the image size */
				height += 200;
				axis_default_space += 100;
				axis_initial_cursor += 50;
				break;
			case 's': /* Listen to the specified socket */
				fprintf(stderr,"[+] Starting Picviz daemon\n");
				fprintf(stderr,"[+] Listen to %s\n", optarg);
				listen_sock = optarg;
				break;
			case 'T': /* Output plugin (pngcairo, svg, ...) */
				sprintf(plugin_out_str, "libpicvizout%s.so", optarg);
				plugin_in = 1;
				break;
			case 't': /* Template file to use when run in daemon mode */
				template = optarg;
				break;
			case 'v':
			case 'V':
				picviz_debug(PICVIZ_DEBUG_NOTICE, PICVIZ_AREA_CORE, "pcv version %s\n", PCV_VERSION);
				exit(0);
				break;
			case 'W': /* Argument for the engine */
				if ( ! strcmp("pcre", optarg)) {
					engine.use_pcre = 1;
					break;
				}
				if ( ! strcmp("string_algo_basic", optarg)) {
					engine.string_algo = 0;
					break;
				}
				fprintf(stderr,"ERROR: unrecognized -W%s.\n", optarg);
				fprintf(stderr,"ERROR: unknown engine parameter!\n");
				help(argv[0]);
				break;
			default: /* '?' */
				fprintf(stderr,"ERROR: arguments parsing!\n");
				help(argv[0]);
		}
	}

	picviz_init(&argc, argv);

	if (!plugin_in) {
		fprintf(stderr,"ERROR: plugin required!\n");
		help(argv[0]);
	}

	if (listen_sock) {

		filter = concat_argv(&argv[optind]);
		engine.real_time = 1;
		if (!template) {
			fprintf(stderr, "No template given. Add '-t template.pcv'.\n");
			exit(1);
		}

		write_pid(engine.pid_file);

		/* We create the image template with its associated filter */
		image = (PicvizImage *)pcv_parse(template, filter);
		if (!image) {
			fprintf(stderr, "Cannot create image. Exiting!\n");
			exit(1);
		}
		picviz_render_image(image); /* We render our template */
		picviz_plugin_load(PICVIZ_PLUGIN_OUTPUT, plugin_out_str, image, plugin_arg);

		picviz_fifo_data_read(image, listen_sock, image_write_callback);

	} else {
		filter = concat_argv(&argv[optind+1]);
		if (optind >= argc) {
			fprintf(stderr, "File to parse missing\n");
			exit(EXIT_FAILURE);
		}

		filename = argv[optind];

		engine.axis_default_space += axis_default_space;
		engine.initial_axis_x_cursor += axis_initial_cursor;
		engine.image_height += height;
		image = (struct pcimage_t *)pcv_parse(filename, filter);
		if (!image) {
			fprintf(stderr, "Cannot parse image %s\n", argv[optind]);
			exit(EXIT_FAILURE);
		}

		/* if (debug) { */
		/* 	picviz_image_debug_printall(image); */
		/* 	goto out; */
		/* } */

		if (plugin_ren_in) {
			picviz_plugin_load(PICVIZ_PLUGIN_RENDER, plugin_ren_str, image, plugin_arg);
		}
		picviz_plugin_load(PICVIZ_PLUGIN_OUTPUT, plugin_out_str, image, plugin_arg);

	}


out:
	picviz_image_destroy(image);

	exit(EXIT_SUCCESS);
}
